zipkin-collector/rabbitmq/src/test/java/zipkin/collector/rabbitmq/LazyRabbitMQCollector.java

```diff
@@ -0,0 +1,122 @@
+ * the License.
+ */
+package zipkin.collector.rabbitmq;
+
+import com.rabbitmq.client.Channel;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.TimeoutException;
+import org.junit.AssumptionViolatedException;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.HostPortWaitStrategy;
+import zipkin.Component;
+import zipkin.collector.InMemoryCollectorMetrics;
+import zipkin.internal.LazyCloseable;
+import zipkin.storage.InMemoryStorage;
+
+class LazyRabbitMQCollector extends LazyCloseable<RabbitMQCollector>
+  implements TestRule {
+  static final int RABBIT_PORT = 5672;
+
+  final InMemoryStorage storage = new InMemoryStorage();
+  final InMemoryCollectorMetrics metrics = new InMemoryCollectorMetrics();
+  final InMemoryCollectorMetrics rabbitmqMetrics = metrics.forTransport("rabbitmq");
+
+  final String image;
+  GenericContainer container;
+
+  LazyRabbitMQCollector(String image) {
+    this.image = image;
+  }
+
+  @Override protected RabbitMQCollector compute() {
+    try {
+      container = new GenericContainer(image)
+        .withExposedPorts(RABBIT_PORT)
+        .waitingFor(new HostPortWaitStrategy());
+      container.start();
+      System.out.println("Starting docker image " + image);
+    } catch (RuntimeException e) {
+      // Ignore
+    }
+
+    RabbitMQCollector result = computeCollectorBuilder().build();
+    result.start();
+
+    Component.CheckResult check = result.check();
+    if (!check.ok) {
+      throw new AssumptionViolatedException(check.exception.getMessage(), check.exception);
+    }
+    return result;
+  }
+
+  RabbitMQCollector.Builder computeCollectorBuilder() {
+    return RabbitMQCollector.builder()
+      .storage(storage)
+      .metrics(metrics)
+      .queue("zipkin-test")
+      .addresses(Arrays.asList(address()));
+  }
+
+  String address() {
+    if (container != null && container.isRunning()) {
+      return String.format("%s:%d",
+        container.getContainerIpAddress(),
+        container.getMappedPort(RABBIT_PORT)
+      );
+    } else {
+      // Use localhost if we failed to start a container (i.e. Docker is not available)
+      return "localhost:" + RABBIT_PORT;
+    }
+  }
+
+  void publish(byte[] message) throws IOException, TimeoutException {
+    RabbitMQCollector thing = get();
+    Channel channel = get().connection.get().createChannel();
+    try {
+      channel.basicPublish("", thing.queue, null, message);
+    } finally {
+      channel.close();
+    }
+  }
+
+  @Override public void close() throws IOException {
+    try {
+      RabbitMQCollector collector = maybeNull();
+      if (collector != null) collector.close();
+    } finally {
+      if (container != null) {
+        System.out.println("Stopping docker image " + image);
+        container.stop();
```

Comment:
`compute` runs once, lazily, but you call `close` many times (after every JUnit statement). This is why the tests are failing

Reply:
@adriancole

Reply:
hmm why would this pass locally!

Reply:
since this is a class rule, it shouldn't run for each test, right?

The error looks like it is failing even before tests are run...

```
Running zipkin.collector.rabbitmq.ITRabbitMQCollector
        ℹ︎ Checking the system...
        ✔ Docker version is newer than 1.6.0
        ✔ Docker environment has more than 2GB free
        ✔ File should be mountable
        ✔ Exposed port is accessible
Starting docker image rabbitmq:3.6-alpine
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 5.925 s <<< FAILURE! - in zipkin.collector.rabbitmq.ITRabbitMQCollector
[ERROR] zipkin.collector.rabbitmq.ITRabbitMQCollector  Time elapsed: 5.925 s  <<< ERROR!
java.lang.IllegalStateException: Unable to establish connection to RabbitMQ server
Caused by: java.io.IOException
Caused by: com.rabbitmq.client.ShutdownSignalException: connection error
Caused by: java.io.EOFException
```

Reply:
@bsideup 
For our other tests run with test containers, I see the following output before the `Checking the system...` sequence:

```
[INFO] Running zipkin.storage.mysql.ITMySQLStorage
07:07:47.617 [main] INFO  o.t.d.EnvironmentAndSystemPropertyClientProviderStrategy - Found docker client settings from environment
07:07:47.633 [main] INFO  o.t.d.DockerClientProviderStrategy - Found Docker environment with Environment variables, system properties and defaults. Resolved: 
    dockerHost=unix:///var/run/docker.sock
    apiVersion='{UNKNOWN_VERSION}'
    registryUrl='https://index.docker.io/v1/'
    registryUsername='travis'
    registryPassword='null'
    registryEmail='null'
    dockerConfig='DefaultDockerClientConfig[dockerHost=unix:///var/run/docker.sock,registryUsername=travis,registryPassword=<null>,registryEmail=<null>,registryUrl=https://index.docker.io/v1/,dockerConfig=/home/travis/.docker,sslConfig=<null>,apiVersion={UNKNOWN_VERSION}]'
07:07:47.638 [main] INFO  o.testcontainers.DockerClientFactory - Docker host IP address is localhost
07:07:47.780 [main] INFO  o.testcontainers.DockerClientFactory - Connected to docker: 
  Server Version: 17.03.1-ce
  API Version: 1.27
  Operating System: Ubuntu 14.04.5 LTS
  Total Memory: 7479 MB
        ℹ︎ Checking the system...
        ✔ Docker version is newer than 1.6.0
        ✔ Docker environment has more than 2GB free
        ✔ File should be mountable
        ✔ Exposed port is accessible
```

Note it is entirely missing for the RabbitMQ tests case (as you can see from the output Adrian posted above), which leads me to believe that a different Docker discovery strategy is being used, but looking at the code for the tests I can't figure out why this would be different.